### PR TITLE
WIP feat/self supervised pretraining

### PR DESCRIPTION
**Pull request by kkontoudi**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*

**What kind of change does this PR introduce?**
Introduces self supervised pre-training. I am opening it to discuss implementation since https://github.com/dreamquark-ai/tabnet/pull/220 was opened on the same topic.

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**

**Closing issues**

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: pytorch_tabnet/tab_network.py [hunk 1]
```diff
@@ -73,11 +73,11 @@ def __init__(self, input_dim, output_dim,
             Avoid log(0), this should be kept very low
         - mask_type: str
             Either "sparsemax" or "entmax" : this is the masking function to use
+        - prior_fn: callable
```
Comment (Optimox) [inline pytorch_tabnet/tab_network.py:76]: Out of curiosity what kind of other priors would you like to use?

File: pytorch_tabnet/tab_network.py [hunk 2]
```diff
@@ -191,6 +179,163 @@ def forward_masks(self, x):
...
+
+        return out, M_loss
+
+    def forward_masks(self, x):
+        self.encoder.forward_masks(x)
+
+
+class TabNetDecoder(torch.nn.Module):
+    def __init__(self, input_dim, output_dim,
+                 n_hidden=8, # size of the hidden layer between the FT and the FC
```
Comment (Optimox) [inline pytorch_tabnet/tab_network.py:260]: hmm, it could indeed be a good idea to give an extra parameter for the hidden dimension of the decoder

File: pytorch_tabnet/tab_network.py (hunk 1 above)
Comment (kkontoudi) [reply pytorch_tabnet/tab_network.py:76]: Well, maybe I went a bit too far by making the ``prior_fn`` an argument here.
I just need it to be a class attribute so that we can change it on every epoch start. This is needed as at every epoch we need to mask different values of the inputs and the priors need to be adjusted accordingly as explained in section 4 of the paper.

**Comment by Optimox** [comment]
Thanks @kkontoudi for opening this PR and sharing your work.
It was helpful to advance in this subject. Hope you'll share again and that this time we will be able to merge.
